hibernate-core/src/main/java/org/hibernate/stat/internal/StatsConcurrentHashMap.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.stat.internal;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Decorates a ConcurrentHashMap implementation to make sure the methods are being
+ * used correctly for the purpose of Hibernate's statistics.
+ * In particular, we do like the semantics of ConcurrentHashMap#computeIfAbsent
+ * but not how it performs.
+ * See also:
+ *  - http://jsr166-concurrency.10961.n7.nabble.com/Re-ConcurrentHashMap-computeIfAbsent-td11687.html
+ *  - https://hibernate.atlassian.net/browse/HHH-13527
+ *
+ * @author Sanne Grinovero
+ */
+final class StatsConcurrentHashMap<V> {
```

Comment:
The name ```StatsConcurrentHashMap``` makes it sounds like it is a ```HashMap```, but it really isn't (since the class does not extend ```HashMap```).

Reply:
good point. How should I call it?

Reply:
I've been trying to think of a good name, but having trouble.

I would prefer to have a separate class for each type of statistic, to be able to encapsulate a description and the computation for the particular type. I suppose it wouldn't perform as well, but I think it would be a lot more clear.

Should I implement my idea?

Reply:
I wouldn't do that. The only reason I created a new container class is to have them behave consistently: I just needed to ensure all statistics usage would avoid using computeIfAbsent but rather use the alternative which always starts with an optimistic get.

Having separate container implementations would not help with consistency.